### PR TITLE
Viewer/accept zero exposuretime

### PR DESCRIPTION
**Pull request by Fixstars-momoko**
Current arv-viewer accepts `>= 1.0`  `ExposureTime`, while GenICam defines its value is `>= 0`. 

This disables the slider when the minimum of `ExposureTime` is less than 0, so I shifted the value +1.0 so that it makes log10(min) more than or equal to 0.

Reference: https://www.emva.org/wp-content/uploads/GenICam_SFNC_v2_7.pdf

### CODE REVIEW

**Comment by EmmanuelP** [comment]
It would be clearer and simpler to clamp the minimum exposure value returned by arv_camera_get_exposure_time_bounds here: https://github.com/AravisProject/aravis/blob/73a4ec46e47c2b9d20f77491ff2e71e23141618b/viewer/arvviewer.c#L729

**Comment by EmmanuelP** [comment]
It would event bette to get back to a linear progression if minimum exposure time is <= 0

**Comment by EmmanuelP** [comment]
As a bonus, you may also use the Representation property of ExposureTime.

It is explained page 33 of the Genicam specification: https://www.emva.org/wp-content/uploads/GenICam_Standard_v2_1_1.pdf

**Comment by Fixstars-momoko** [comment]
@EmmanuelP Hi, thanks for the feedback and the info of Representation https://github.com/AravisProject/aravis/pull/821#issuecomment-1727038841

I have updated gtk widget as follows:

| Representation | Linear  | Logarithmic | PureNumber | Undefined | default /other |
|----------------|---------|-------------|------------|---------|---------|
| edit box | enabled | enabled | enabled | enabled | disabled |
| slider | enabled | enabled | disabled | enabled | disabled |
| slider-scale | min-max| 0-1 | N/A | min-max | N/A | 

I was not sure in the case of Undefined (-1), so I tentatively enabled both widgets. Please let me know if they are supposed to be disabled.

File: src/arvcamera.c [hunk 1]
```diff
@@ -2427,6 +2427,65 @@ arv_camera_is_binning_available (ArvCamera *camera, GError **error)
...
+ * @camera: a #ArvCamera
+ * @error: a #GError placeholder, %NULL to ignore
+ *
+ * Return: enum ArvGcRepresentation, -1 if not available.
+ *
+ * Since: 
+ */
+
+ArvGcRepresentation
+arv_camera_get_representation_gain (ArvCamera *camera, GError **error)
```
Comment (EmmanuelP) [inline src/arvcamera.c:2441]: Please rename this function to arv_camera_get_gain_representation()
Remove the error placeholder. Errors are for communication problems, not programming errors. get_representation comes directly from the xml data.

File: src/arvcamera.c [hunk 2]
```diff
@@ -2427,6 +2427,65 @@ arv_camera_is_binning_available (ArvCamera *camera, GError **error)
...
+ * @camera: a #ArvCamera
+ * @error: a #GError placeholder, %NULL to ignore
+ *
+ * Return: enum ArvGcRepresentation, -1 if not available.
+ *
+ * Since: 
+ */
+
+ArvGcRepresentation
+arv_camera_get_representation_exposure_time (ArvCamera *camera, GError **error)
```
Comment (EmmanuelP) [inline src/arvcamera.c:2469]: Rename this function to arv_camera_get_exposure_time_representation()
Remove the error placeholder

File: src/arvcamera.c [hunk 3]
```diff
@@ -2427,6 +2427,65 @@ arv_camera_is_binning_available (ArvCamera *camera, GError **error)
...
+{
+	ArvCameraPrivate *priv = arv_camera_get_instance_private (camera);
+
+	g_return_val_if_fail (ARV_IS_CAMERA (camera), FALSE);
```
Comment (EmmanuelP) [inline src/arvcamera.c:2473]: Type mismatch, return REPRESENTATION_UNDEFINED instead, and update the doc accordingly

File: src/arvcamera.c [hunk 4]
```diff
@@ -2427,6 +2427,65 @@ arv_camera_is_binning_available (ArvCamera *camera, GError **error)
...
+ *
+ * Since: 
+ */
+
+ArvGcRepresentation
+arv_camera_get_representation_gain (ArvCamera *camera, GError **error)
+{
+	ArvCameraPrivate *priv = arv_camera_get_instance_private (camera);
+
+	g_return_val_if_fail (ARV_IS_CAMERA (camera), FALSE);
```
Comment (EmmanuelP) [inline src/arvcamera.c:2445]: Type mismatch, return REPRESENTATION_UNDEFINED instead, and update the doc accordingly

File: src/arvdevice.c [hunk 5]
```diff
@@ -393,6 +393,35 @@ _get_feature (ArvDevice *device, GType node_type, const char *feature, GError **
...
+ * @device: a #ArvDevice
+ * @feature: feature name
+ * @error: a #GError placeholder, %NULL to ignore
+ *
+ * Return: enum ArvGcRepresentation, -1 if not available.
+ *
+ * Since: 
+ */
+
+ArvGcRepresentation
```
Comment (EmmanuelP) [inline src/arvdevice.c:407]: Remove the error placeholder

File: src/arvdevice.c [hunk 6]
```diff
@@ -393,6 +393,35 @@ _get_feature (ArvDevice *device, GType node_type, const char *feature, GError **
...
+arv_device_get_representation (ArvDevice *device, const char *feature, GError **error)
+{
+	ArvGcNode* node;
+
+	g_return_val_if_fail (ARV_IS_DEVICE (device), FALSE);
+	g_return_val_if_fail (feature != NULL, FALSE);
```
Comment (EmmanuelP) [inline src/arvdevice.c:413]: Type mismatch, return REPRESENTATION_UNDEFINED instead, and update the doc accordingly

File: viewer/arvviewer.c [hunk 7]
```diff
@@ -534,8 +543,10 @@ black_level_spin_cb (GtkSpinButton *spin_button, ArvViewer *viewer)
 static void
 exposure_scale_cb (GtkRange *range, ArvViewer *viewer)
 {
-	double log_exposure = gtk_range_get_value (range);
-	double exposure = arv_viewer_value_from_log (log_exposure, viewer->exposure_min, viewer->exposure_max);
+	double scaled_exposure = gtk_range_get_value (range);
+	double exposure = viewer->exposure_time_representation == ARV_GC_REPRESENTATION_LOGARITHMIC ?
+		arv_viewer_value_from_log (scaled_exposure, viewer->exposure_min, viewer->exposure_max) : scaled_exposure;
+	printf("scaled_exposure=%f exposure=%f\n", scaled_exposure, exposure);
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:549]: Debug leftover, to be removed

File: viewer/arvviewer.c [hunk 8]
```diff
@@ -731,16 +749,16 @@ set_camera_widgets(ArvViewer *viewer)
...
 	gtk_spin_button_set_increments (GTK_SPIN_BUTTON (viewer->gain_spin_button), 1, 10);
 
 	arv_camera_get_black_level_bounds (viewer->camera, &black_level_min, &black_level_max, NULL);
 	gtk_spin_button_set_range (GTK_SPIN_BUTTON (viewer->black_level_spin_button), black_level_min, black_level_max);
 	gtk_spin_button_set_increments (GTK_SPIN_BUTTON (viewer->black_level_spin_button), 1, 10);
 
-	gtk_range_set_range (GTK_RANGE (viewer->exposure_hscale), 0.0, 1.0);
-	gtk_range_set_range (GTK_RANGE (viewer->gain_hscale), gain_min, gain_max);
+	// gtk_range_set_range (GTK_RANGE (viewer->exposure_hscale), 0.0, 1.0);
+	// gtk_range_set_range (GTK_RANGE (viewer->gain_hscale), viewer->gain_min, viewer->gain_max);
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:761]: Don't leave commented code

File: viewer/arvviewer.c [hunk 9]
```diff
@@ -142,8 +142,12 @@ struct  _ArvViewer {
 	gboolean flip_vertical;
 	gboolean flip_horizontal;
 
+	double gain_min, gain_max;
 	double exposure_min, exposure_max;
 
+	gint gain_representation;
+	gint exposure_time_representation;
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:149]: Wrong type, use ArvGcRepresentation

File: viewer/arvviewer.c [hunk 10]
```diff
@@ -749,17 +767,66 @@ set_camera_widgets(ArvViewer *viewer)
 	string = g_strdup_printf ("%g", arv_camera_get_frame_rate (viewer->camera, NULL));
 	gtk_entry_set_text (GTK_ENTRY (viewer->frame_rate_entry), string);
 
+	// enables gtk widget based on feature representation
 	is_gain_available = arv_camera_is_gain_available (viewer->camera, NULL);
-	gtk_widget_set_sensitive (viewer->gain_hscale, is_gain_available);
-	gtk_widget_set_sensitive (viewer->gain_spin_button, is_gain_available);
+	if (is_gain_available){
+		viewer->gain_representation = (gint)(arv_camera_get_representation_gain(viewer->camera, NULL));
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:773]: Remove the cast to gint

File: viewer/arvviewer.c [hunk 11]
```diff
@@ -749,17 +767,66 @@ set_camera_widgets(ArvViewer *viewer)
...
 
 	is_black_level_available = arv_camera_is_black_level_available (viewer->camera, NULL);
 	gtk_widget_set_sensitive (viewer->black_level_hscale, is_black_level_available);
 	gtk_widget_set_sensitive (viewer->black_level_spin_button, is_black_level_available);
 
 	is_exposure_available = arv_camera_is_exposure_time_available (viewer->camera, NULL);
-	gtk_widget_set_sensitive (viewer->exposure_hscale, is_exposure_available);
-	gtk_widget_set_sensitive (viewer->exposure_spin_button, is_exposure_available);
+	if (is_exposure_available){
+		viewer->exposure_time_representation = (gint)(arv_camera_get_representation_exposure_time(viewer->camera, NULL));
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:805]: Remove the cast to gint

File: viewer/arvviewer.c [hunk 12]
```diff
@@ -749,17 +767,66 @@ set_camera_widgets(ArvViewer *viewer)
 	string = g_strdup_printf ("%g", arv_camera_get_frame_rate (viewer->camera, NULL));
 	gtk_entry_set_text (GTK_ENTRY (viewer->frame_rate_entry), string);
 
+	// enables gtk widget based on feature representation
```
Comment (EmmanuelP) [inline viewer/arvviewer.c:770]: Please remove the comment, the code is self explanatory

**Review by EmmanuelP — CHANGES_REQUESTED** [review]
Thanks for the pull request. The comments are mainly coding style issues. Please fix them an I'll merge the changes.

File: src/arvcamera.c (hunk 1 above)
Comment (Fixstars-momoko) [reply src/arvcamera.c:2441]: updated in b9c1305b92d0322ec1c7f1ebef44a47d78db6a99

File: src/arvcamera.c (hunk 2 above)
Comment (Fixstars-momoko) [reply src/arvcamera.c:2469]: updated in b9c1305b92d0322ec1c7f1ebef44a47d78db6a99

File: src/arvdevice.c (hunk 5 above)
Comment (Fixstars-momoko) [reply src/arvdevice.c:407]: updated in b9c1305b92d0322ec1c7f1ebef44a47d78db6a99

File: src/arvcamera.c (hunk 3 above)
Comment (Fixstars-momoko) [reply src/arvcamera.c]: Updated in 17a2454

File: src/arvcamera.c (hunk 4 above)
Comment (Fixstars-momoko) [reply src/arvcamera.c]: Updated in 17a2454

File: src/arvdevice.c (hunk 6 above)
Comment (Fixstars-momoko) [reply src/arvdevice.c:412]: updated in dbf13af

File: viewer/arvviewer.c (hunk 12 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c:770]: updated in 1fd760c

File: viewer/arvviewer.c (hunk 8 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c:761]: updated in 1fd760c

File: viewer/arvviewer.c (hunk 7 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c:549]: updated in 1fd760c

File: viewer/arvviewer.c (hunk 11 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c]: updated in c5ccba1

File: viewer/arvviewer.c (hunk 10 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c]: updated in c5ccba1

File: viewer/arvviewer.c (hunk 9 above)
Comment (Fixstars-momoko) [reply viewer/arvviewer.c:149]: updated in c5ccba1

**Comment by EmmanuelP** [comment]
Thanks a lot. I have merged your changes with some minor fixes.

I may revert the widget selection based on representation though, at least for exposure time, as it looks like most of the time it is defined as linear, while logarithmic makes more sense.

**Pull request closed without merging**